vertx-auth-mtls/src/main/java/io/vertx/ext/auth/mtls/impl/MutualTlsAuthenticationProvider.java

```
@@ -0,0 +1,93 @@
+ * both server and client exchange certificates and validates each other's certificates.
+ */
+public class MutualTlsAuthenticationProvider implements AuthenticationProvider
+{
+    private final CertificateValidator certificateValidator;
+    private final CertificateIdentityExtractor identityExtractor;
+    private final CertificateIdentityValidator identityValidator;
+
+    public MutualTlsAuthenticationProvider(CertificateValidator certificateValidator,
+                                           CertificateIdentityExtractor identityExtractor,
+                                           CertificateIdentityValidator identityValidator)
+    {
+        this.certificateValidator = certificateValidator;
+        this.identityExtractor = identityExtractor;
+        this.identityValidator = identityValidator;
+    }
+
+    @Override
+    public Future<User> authenticate(Credentials credentials)
+    {
+        if (!(credentials instanceof CertificateCredentials))
+        {
+            return Future.failedFuture("CertificateCredentials expected for mTLS authentication");
+        }
+
+        CertificateCredentials certificateCredentials = (CertificateCredentials) credentials;
+        String identity;
+        try
+        {
+            certificateValidator.isValidCertificate(certificateCredentials);
+            identity = identityExtractor.identity(certificateCredentials.certificateChain().toArray(new Certificate[0]));
+        }
+        catch (Exception e)
+        {
+            return Future.failedFuture("Invalid certificate passed");
+        }
+
+        if (identity == null || identity.isEmpty())
+        {
+            return Future.failedFuture("Could not extract identity from certificate");
+        }
+        if (!identityValidator.isValidIdentity(identity))
+        {
+            return Future.failedFuture("Certificate identity is not a valid identity");
+        }
```

You can merge the validation code with the try-catch block above, once you made the interface to throw exception, instead of returning boolean.